src/script/cpatch.py

```diff
@@ -0,0 +1,727 @@
+    ):
+
+        for cur, dirs, files in os.walk(start_path):
+            cur = pathlib.Path(cur)
+            # skip 'tests' directories in both tar and walk
+            if exclude_dirs:
+                _exclude_dirs(dirs, exclude_dirs)
+            for sdir in dirs:
+                ddir = cur / sdir
+                tar.add(ddir, recursive=False)
+            for sfile in files:
+                if exclude_file_suffixes and sfile.endswith(
+                    exclude_file_suffixes
+                ):
+                    continue
+                dfile = cur / sfile
+                tar.add(dfile, recursive=False)
+
+    def _copy_binary(self, src_path, dst_path):
+        log.debug("binary: %s -> %s", src_path, dst_path)
+        if self._ctx.strip_binaries:
+            log.debug("copy and strip: %s", dst_path)
+            shutil.copy2(src_path, dst_path)
+            _run(["strip", str(dst_path)]).check_returncode()
+            return
+        log.debug("hard linking: %s", dst_path)
+        try:
+            os.unlink(dst_path)
+        except FileNotFoundError:
+            pass
+        os.link(src_path, dst_path)
+
+    def _bins_and_libs(self, prefix, bin_patterns, lib_patterns):
+        out = []
+
+        bin_src = self._ctx.build_dir / "bin"
+        bin_dst = self._workdir / f"{prefix}_bin"
+        bin_dst.mkdir(parents=True, exist_ok=True)
+        for path in bin_src.iterdir():
+            if any(path.match(m) for m in bin_patterns):
+                self._copy_binary(path, bin_dst / path.name)
+        out.append(f"ADD {prefix}_bin /usr/bin")
+
+        lib_src = self._ctx.build_dir / "lib"
+        lib_dst = self._workdir / f"{prefix}_lib"
+        lib_dst.mkdir(parents=True, exist_ok=True)
+        for path in lib_src.iterdir():
+            if any(path.match(m) for m in lib_patterns):
+                self._copy_binary(path, lib_dst / path.name)
+        out.append(f"ADD {prefix}_lib /usr/lib64")
+
+        return out
+
+    def _conditional_libs(self, src_dir, name, destination, lib_patterns):
+        lib_src = self._ctx.build_dir / src_dir
+        lib_dst = self._workdir / name
+        lib_dst.mkdir(parents=True, exist_ok=True)
+        try:
+            for path in lib_src.iterdir():
+                if any(path.match(m) for m in lib_patterns):
+                    self._copy_binary(path, lib_dst / path.name)
+        except FileNotFoundError as err:
+            log.warning("skipping lib %s: %s", name, err)
+        return f"ADD {name} {destination}"
+
+    def _py_site_packages(self):
+        """Return the correct python site packages dir for the image."""
+        if self._cached_py_site_packages is not None:
+            return self._cached_py_site_packages
+        # use the container image to probe for the correct python site-packages dir
+        valid_site_packages = [
+            "/usr/lib/python3.8/site-packages",
+            "/usr/lib/python3.6/site-packages",
```

Comment:
do we want to preemptively add the default python version for centos 9 since that might be used for builds after we drop centos 8 support in the squid cycle? I want to say it's 3.9.

Reply:
That's an interesting thought, we should do something about it once ceph starts producing base images using centos9stream or whatnot as a base.